[compiler-rt][ARM] Optimized mulsf3 and divsf3

statham-arm opened the pull request:
This commit adds optimized assembly versions of single-precision float multiplication and division. Both functions are implemented in a style that can be assembled as either of Arm and Thumb2; for multiplication, a separate implementation is provided for Thumb1. Also, extensive new tests are added for multiplication and division.

These implementations can be removed from the build by defining the cmake variable COMPILER_RT_ARM_OPTIMIZED_FP=OFF.

Outlying parts of the functionality which are not on the fast path, such as NaN handling and underflow, are handled in helper functions written in C. These can be shared between the Arm/Thumb2 and Thumb1 implementations, and also reused by other optimized assembly functions we hope to add in future.

statham-arm committed:
[compiler-rt][ARM] Optimized mulsf3 and divsf3

This commit adds optimized assembly versions of single-precision float
multiplication and division. Both functions are implemented in a style
that can be assembled as either of Arm and Thumb2; for multiplication,
a separate implementation is provided for Thumb1. Also, extensive new
tests are added for multiplication and division.

These implementations can be removed from the build by defining the
cmake variable COMPILER_RT_ARM_OPTIMIZED_FP=OFF.

Outlying parts of the functionality which are not on the fast path,
such as NaN handling and underflow, are handled in helper functions
written in C. These can be shared between the Arm/Thumb2 and Thumb1
implementations, and also reused by other optimized assembly functions
we hope to add in future.

aykevl commented on compiler-rt/lib/builtins/CMakeLists.txt:
```diff
@@ -422,6 +422,22 @@ set(arm_or_thumb2_base_SOURCES
 )
 
+option(COMPILER_RT_ARM_OPTIMIZED_FP
+  "On 32-bit Arm, use optimized assembly implementations of FP arithmetic" ON)
```
I believe this is a code size vs speed tradeoff, right?
I think it would be a good idea to say that explicitly. (And IMHO if the new assembly routines are both smaller and faster they should just be replaced instead of having two options).

aykevl reviewed (commented):
Can't give much of a review here, but superficially this looks fine to me.

compnerd commented on compiler-rt/lib/builtins/arm/thumb1/mulsf3.S:
```diff
@@ -0,0 +1,251 @@
+
+DEFINE_AEABI_FUNCTION_ALIAS(__aeabi_fmul, __mulsf3)
+
+DEFINE_COMPILERRT_FUNCTION(__mulsf3)
```
Given the previous `.thumb`, I think that we should use `DEFINE_COMPILERRT_THUMB_FUNCTION` instead.

compnerd commented on compiler-rt/lib/builtins/arm/thumb1/mulsf3.S:
```diff
@@ -0,0 +1,251 @@
+  LSLS    r3, r2, #23
+  ADDS    r0, r0, r3    // put on the biased exponent
+
+  BL      __funder
```
Should this be `SYMBOL_NAME(__compiler_rt_funder)`?

compnerd commented on compiler-rt/lib/builtins/arm/thumb1/mulsf3.S:
```diff
@@ -0,0 +1,251 @@
+LOCAL_LABEL(denorm):
+  PUSH    {r0,r1,r2,r3}
+  MOV     r0, sp
+  BL      __fnorm2
```
Should this be `SYMBOL_NAME(__compiler_rt_fnorm2)`?

statham-arm replied:
How embarrassing. Not only that, but I hadn't actually added the helper functions to the build in the Thumb1 case. Apparently forgot to re-test both architectures before pushing! Fixed now.

compnerd commented on compiler-rt/lib/builtins/arm/thumb1/mulsf3.S:
```diff
@@ -0,0 +1,251 @@
+  // propagates an appropriate NaN to the output, dealing with the special
+  // cases of signalling/quiet NaNs.
+LOCAL_LABEL(nan):
+  BL      __fnan2
```
Should this be `SYMBOL_NAME(__compiler_rt_fnan2)`?

compnerd commented on compiler-rt/lib/builtins/arm/divsf3.S:
```diff
@@ -0,0 +1,608 @@
+
+*/
+
+  .p2align 2  // make sure we start on a 32-bit boundary, even in Thumb
```
I think that changing this to `4-byte boundary` is better than `32-bit boundary` as it can be confusing when scanning over the comment and code.

statham-arm replied:
Done.

compnerd commented on compiler-rt/lib/builtins/arm/fnan2.c:
```diff
@@ -0,0 +1,38 @@
+    return a | 0x00400000; //   if so, return it with the quiet bit set
+  if (badj > 0xff800000)   // b is a signalling NaN?
+    return b | 0x00400000; //   if so, return it with the quiet bit set
+  if (aadj < 0x00800000)   // a is a quiet NaN?
+    return a;              // if so, return it
+  else                     // expect (badj < 0x00800000)
+    return b;              // in that case b must be a quiet NaN
```
I think that this should be either of the following:

```
return (aadj < 0x00800000) ? a : b;
```

or

```
if (aadj < 0x00800000)
  return a;
return b;
```

statham-arm committed:
Fix the Thumb1 build which I forgot to test

statham-arm committed:
Use DEFINE_COMPILERRT_THUMB_FUNCTION in Thumb1

statham-arm committed:
Tweak final return in fnan2 as suggested

statham-arm committed:
Clarify comment about 4-byte boundary

statham-arm committed:
Lowercase instruction mnemonics and shifter operands

This was Petr Hosek's comment on #154093, but if we're doing that, we
should do it consistently.

statham-arm committed:
Mention size/speed tradeoff in the cmake option help

statham-arm committed:
Update build and test setup

Now we should only test the extra NaN faithfulness in cases where it's
provided by the library. Also tweaked the cmake setup to make it
easier to add more assembly files later. Plus a missing piece of
comment in fnan2.c.

statham-arm commented on compiler-rt/lib/builtins/CMakeLists.txt:
```diff
@@ -422,6 +422,24 @@ set(arm_or_thumb2_base_SOURCES
+    arm/mulsf3.S
+    arm/divsf3.S)
+  set_source_files_properties(${assembly_files}
+    PROPERTIES COMPILE_OPTIONS "-Wa,-mimplicit-it=always")
```
I'm still very confused about _how_ to configure a test build of compiler-rt which will successfully assemble the Arm builtins without being a compiler driver that speaks `-Wa,`. You seem to be saying here that there's some other way to configure cmake to invoke both cpp and as? But I don't know what it is; surely CMAKE_ASM_COMPILER would need to be set to a single command, and I can't think what command would work.

> In fact, I think that `-mimplicit-it=always` can just be passed without the `-Wa,`

Huh, apparently that works in clang! TIL. Doesn't work in my nearest gcc, though – that still demands `-Wa,`. But I suppose at least that gives me a way to test a piece of cmake that checks which flag works.

statham-arm replied:
OK, now I've put in a check. It turned out that no cmake command higher-level than the general-purpose `try_compile` would handle assembly language -- all the things like `check_compiler_flag()` had a list of supported languages not including ASM. So I had to write my own helper function.

I still haven't figured out how to test it against anything that isn't a compiler driver. But I'm testing for `-mimplicit-it` first, which means that compiling with clang the first test passes, and with gcc the first test fails and we fall back to `-Wa,-mimplicit-it`, so I've at least been able to check that these tests are doing _something_ nontrivial.

statham-arm committed:
Check for the right spelling of the implicit-it option

statham-arm committed:
Fix build failure on every non-Arm platform

Oops! I forgot that COMPILER_RT_ARM_OPTIMIZED_FP defaulted to on, so
that this check would run even for builds that never even knew about it.

compnerd reviewed (approved):
Thanks!
